wip/memory-interface.md

```diff
@@ -0,0 +1,137 @@
+# Rust Memory Interface
+
+**Note:** This document is not normative nor endorsed by the UCG WG. Its purpose is to be the basis for discussion and to set down some key terminology.
+
+The purpose of this document is to describe the interface between a Rust program and memory.
+This interface is a key part of the Rust Abstract Machine: it lets us separate concerns by splitting the Machine (i.e., its specification) into two pieces, connected by this well-defined interface:
+* The *expression/statement semantics* of Rust boils down to explaining which "memroy events" (calls to the memory interface) happen in which order. This part of the specification is *pure* in the sense that it has no "state": everything that needs to be remembered from one expression evaluation to the next is communicated through memory.
+* The Rust *memory model* explains which interactions with the memory are legal (the others are UB), and which values can be returned by reads.
+
+The interface shown below is also opinionated in several ways.
+It is not intended to be able to support *any imaginable* memory model, but rather start the process of reducing the design space of what we consider a "reasonable" memory model for Rust.
+For example, it explicitly acknowledges that pointers are not just integers and that uninitialized memory is special (both are true for C and C++ as well but you have to read the standard very careful, and consult [non-normative defect report responses](http://www.open-std.org/jtc1/sc22/wg14/www/docs/dr_260.htm), to see this).
```

Comment:
I thought defect reports were normative, even if the conclusion of the report is "nothing has to be done".

Reply:
From what I was told, they are normative for the standard they were reported against, but when reading C18 I don't have to read all defect reports of all prior versions. That would be ludicrous.

C18 contains the exact same wording as C11 for indeterminate values.

I can still remove the "non-normative" part of the sentence, it doesn't really add anything.

Reply:
> From what I was told, they are normative for the standard they were reported against, but when reading C18 I don't have to read all defect reports of all prior versions. That would be ludicrous.

Sounds correct to me. The changes proposed in defect reports are automatically merged into the standard draft, so they become part of the next standard version automatically. That particular report is a bit weird, in that its conclusion is that the defect isn't a defect because the standard is clear and therefore nothing must be done. If people have to fill a defect report, and reference it everywhere, the standard is obviously not clear..

Reply:
Agreed, I think the C committee's position here is harmful. But alas, that is not something we are going to fix here and now.